### DISCUSSION TITLE
Reason to change GCN weight parameter

### DISCUSSION THREAD

**Discussion by khaled-rahman** [pyg-team/pytorch_geometric#3236] [Q&A] [answered]
Hi, I was checking the weight in the new GCNConv class. I think it was `self.weight = Parameter(torch.Tensor(in_channels, out_channels))` previously. Probably, this has been changed in the recent version with a linear layer. Is there any specific reason for this change or is there any disadvantage of the previous one? It would be great to know.

**Comment by wsad1** [comment, marked as answer]
The GCNConv class uses [torch_geometric.nn.dense.Linear](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.dense.Linear) as it supports lazy initialization, where the parameters are initialized after the first call to forward. `torch.nn.parameter` does not support this.

Why was lazy initialization needed?
1. Users don't need to track in channels, the model will infer it from the first forward call.
2. More importantly, Lazy initialization makes working with heterogenous graphs easier. For example check out the [to_hetero](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html#automatically-converting-gnn-models), which automatically converts a homogenous graph to an heterogenous graph, for this to work GNN layers like GCNConv or SAGEConv need to support lazy initialization.
